crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -2535,6 +2537,53 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         }
     }
 
+    /// Special case for unannotated `cls` and `self` arguments to class methods and instance methods.
+    fn special_first_method_parameter_type(
+        &mut self,
+        parameter: &ast::Parameter,
+    ) -> Option<Type<'db>> {
+        let db = self.db();
+        let scope_id = self.scope();
+        let file = scope_id.file(db);
+        let function_scope = scope_id.scope(db);
+        let method = function_scope.node().as_function()?;
+
+        let parent_scope_id = function_scope.parent()?;
+        let parent_scope = self.index.scope(parent_scope_id);
+        parent_scope.node().as_class()?;
+
+        let method_definition = self.index.expect_single_definition(method);
+        let DefinitionKind::Function(function_definition) = method_definition.kind(db) else {
+            return None;
+        };
+
+        let func_type = infer_definition_types(db, method_definition)
+            .declaration_type(method_definition)
+            .inner_type()
+            .as_function_literal()?;
+
+        let module = parsed_module(db, file).load(db);
+        if function_definition
+            .node(&module)
+            .parameters
+            .index(parameter.name())
+            .is_some_and(|index| index != 0)
+        {
+            return None;
+        }
+
+        if func_type.is_classmethod(db) {
+            // TODO: set the type for `cls` argument
+            return None;
+        } else if func_type.is_staticmethod(db) {
+            return None;
+        }
+
+        let class = nearest_enclosing_class(db, self.index, scope_id).unwrap();
```

Comment:
Nit: Not sure if it's worth it but we already check that the parent is a class on the line above. Could we infer the type directly instead of walking the ancestor chain again?